new-log-viewer/src/services/decoders/JsonlDecoder/index.ts

```diff
@@ -0,0 +1,242 @@
+    }
+
+    /**
+     * Retrieves log event using index then decodes into `DecodeResultType`.
+     *
+     * @param logEventIdx
+     * @return Decoded result.
+     */
+    #getDecodeResult = (logEventIdx: number): DecodeResultType => {
+        let timestamp: number;
+        let message: string;
+        let logLevel: LOG_LEVEL;
+
+        // eslint-disable-next-line no-warning-comments
+        // TODO We could probably optimize this to avoid checking `#invalidLogEventIdxToRawLine` on
+        // every iteration.
+        if (this.#invalidLogEventIdxToRawLine.has(logEventIdx)) {
+            timestamp = INVALID_TIMESTAMP_VALUE;
+            message = `${this.#invalidLogEventIdxToRawLine.get(logEventIdx)}\n`;
+            logLevel = LOG_LEVEL.NONE;
+        } else {
+            // Explicit cast since typescript thinks `#logEvents[filteredIdx]` can be undefined,
+            // but it shouldn't be since the index comes from a class-internal filter.
+            const logEvent = this.#logEvents[logEventIdx] as JsonLogEvent;
+            logLevel = logEvent.level;
+            message = this.#formatter.formatLogEvent(logEvent);
+            timestamp = logEvent.timestamp.valueOf();
+        }
+
+        return [
+            message,
+            timestamp,
+            logLevel,
+            logEventIdx + 1,
+        ];
+    };
+
+    /**
+     * Parses each line from the data array as a JSON object and buffers it internally. If a
+     * line cannot be parsed as a JSON object, an error is logged and the line added to an
+     * invalid map.
+     *
+     * NOTE: The data array(file) is freed after the very first run of this method.
+     */
+    #deserialize () {
+        if (null === this.#dataArray) {
+            return;
+        }
+
+        const text = JsonlDecoder.#textDecoder.decode(this.#dataArray);
+        let beginIdx: number = 0;
+        while (beginIdx < text.length) {
+            const endIdx = text.indexOf("\n", beginIdx);
+            const line = (-1 === endIdx) ?
+                text.substring(beginIdx) :
+                text.substring(beginIdx, endIdx);
+
+            beginIdx = (-1 === endIdx) ?
+                text.length :
+                endIdx + 1;
+
+            this.#parseJson(line);
+        }
+
+        this.#dataArray = null;
+    }
+
+    /**
+     * Parse line into a json log event and buffer internally. If the line contains invalid
+     * json, an entry is added to invalid log event map.
+     *
+     * @param line
+     */
+    #parseJson (line: string) {
+        try {
+            const fields = JSON.parse(line) as JsonValue;
+            if (!isJsonObject(fields)) {
+                throw new Error("Unexpected non-object.");
+            }
+            this.#logEvents.push({
+                fields: fields,
+                level: convertToLogLevelValue(fields[this.#logLevelKey]),
+                timestamp: convertToDayjsTimestamp(fields[this.#timestampKey]),
+            });
```

Comment:
Since we're calling this `push` in both branches, can we use some local variables and then call it once outside the branches?